crates/bevy_ecs/macros/src/component.rs

```diff
@@ -75,100 +75,191 @@ pub fn derive_component(input: TokenStream) -> TokenStream {
 
-    let requires = &attrs.requires;
-    let mut register_required = Vec::with_capacity(attrs.requires.iter().len());
-    let mut register_recursive_requires = Vec::with_capacity(attrs.requires.iter().len());
-    if let Some(requires) = requires {
-        for require in requires {
+    let RelatedComponentsReturn {
+        register_related: register_suggested,
+        register_recursive_related: register_recursive_suggested,
+        docs: suggested_component_docs,
+    } = related_components(&attrs.suggests, Relatedness::Suggested);
+    let RelatedComponentsReturn {
+        register_related: register_included,
+        register_recursive_related: register_recursive_included,
+        docs: included_component_docs,
+    } = related_components(&attrs.includes, Relatedness::Included);
+    let RelatedComponentsReturn {
+        register_related: register_required,
+        register_recursive_related: register_recursive_required,
+        docs: required_component_docs,
+    } = related_components(&attrs.requires, Relatedness::Required);
+
+    let struct_name = &ast.ident;
+    let (impl_generics, type_generics, where_clause) = &ast.generics.split_for_impl();
+
+    // This puts `register_required` before `register_recursive_requires` to ensure that the constructors of _all_ top
+    // level components are initialized first, giving them precedence over recursively defined constructors for the same component type
+    TokenStream::from(quote! {
+        #required_component_docs
+        #included_component_docs
+        #suggested_component_docs
+        impl #impl_generics #bevy_ecs_path::component::Component for #struct_name #type_generics #where_clause {
+            const STORAGE_TYPE: #bevy_ecs_path::component::StorageType = #storage;
+            fn register_related_components(
+                requiree: #bevy_ecs_path::component::ComponentId,
+                components: &mut #bevy_ecs_path::component::Components,
+                storages: &mut #bevy_ecs_path::storage::Storages,
+                required_components: &mut #bevy_ecs_path::component::RelatedComponents,
+                inheritance_depth: u16,
+            ) {
+                #(#register_suggested)*
+                #(#register_recursive_suggested)*
+
+                #(#register_included)*
+                #(#register_recursive_included)*
+
+                #(#register_required)*
+                #(#register_recursive_required)*
+            }
+
+            #[allow(unused_variables)]
+            fn register_component_hooks(hooks: &mut #bevy_ecs_path::component::ComponentHooks) {
+                #on_add
+                #on_insert
+                #on_replace
+                #on_remove
+            }
+        }
+    })
+}
+
+// Because the macro crate cannot access the `bevy_ecs` crate, we need to create our own equivalent.
+// This should be kept in sync with the actual `bevy_ecs` crate!
+enum Relatedness {
```

Comment:
This PR needs enough work, and is "involved" enough that I'm worried about rushing it in this late in the game. This feels like the kind of thing that deserves time to be settled and tested. Additionally, this framing has been on our mind for 3 days. I think it _might_ be the right path forward, but the whole thing is still settling in my head. It feels like we're rushing into this.

Given how the required components conversation is going (in the release post PR), I think we can swing require-only for 0.15, _especially_ if we add the "break-glass "`#[remove_require_risky(Transform)]`". Given that this whole "other related components" effort is motivated by implementing the removal behavior for `include` to nail down the difference between the relationships, and given that this PR does not _yet_ implement that behavior, I think our limited resources / time / risk tolerance should be spent on implementing the removal functionality. That will already be reasonably invasive.

That then gives us a whole cycle to decide what the global picture should look like, ensure the implementation is sound, and actually make use of new tools like `include` ( I don't think we have the time in 0.15 to go through the engine and decide what deserves require vs include).

In the 0.15 release post, we could communicate that we plan on adding additional relationship tools for more flexibility.

Reply:
Yep, I'm on board with that decision. This is involved enough and I have enough to do that I'm not keen on pushing it in last minute.